File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressVerticalGrid/Demos/Delphi/ScrollbarAnnotationsDemo/ScrollbarAnnotationsDemoMain.pas
unit ScrollbarAnnotationsDemoMain;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics,
  Controls, Forms, Dialogs, DemoBasicMain, Menus, cxClasses, cxLookAndFeels,
  ActnList, ImgList, ComCtrls, StdCtrls, dxCore,
  cxGraphics, cxControls, cxLookAndFeelPainters, cxStyles, cxEdit, cxTextEdit,
  cxVGrid, cxDBVGrid, cxInplaceContainer, cxCustomData, dxScrollbarAnnotations;

type
  TScrollbarAnnotationsDemoMainForm = class(TDemoBasicMainForm)
    ActionList1: TActionList;
    actScrollAnnotationsActive: TAction;
    actDoorCount: TAction;
    actPrice: TAction;
    actCylinderCount: TAction;
    actShowErrors: TAction;
    actShowSearchResults: TAction;
    actShowFocusedRow: TAction;
    actCustomAnnotationSettings: TAction;
    cxDBVerticalGrid1: TcxDBVerticalGrid;
    clName: TcxDBEditorRow;
    clModification: TcxDBEditorRow;
    clPrice: TcxDBEditorRow;
    cxDBVerticalGrid1MPGCity: TcxDBEditorRow;
    cxDBVerticalGrid1MPGHighway: TcxDBEditorRow;
    clDoorCount: TcxDBEditorRow;
    clCilinderCount: TcxDBEditorRow;
    cxDBVerticalGrid1Horsepower: TcxDBEditorRow;
    cxDBVerticalGrid1Torque: TcxDBEditorRow;
    cxDBVerticalGrid1TransmissionSpeeds: TcxDBEditorRow;
    cxDBVerticalGrid1TransmissionType: TcxDBEditorRow;
    cxDBVerticalGrid1Description: TcxDBEditorRow;
    cxDBVerticalGrid1DeliveryDate: TcxDBEditorRow;
    Scrollannotations1: TMenuItem;
    Active1: TMenuItem;
    N1: TMenuItem;
    ShowErrors1: TMenuItem;
    ShowSearchResults1: TMenuItem;
    ShowFocusedRow1: TMenuItem;
    N2: TMenuItem;
    Doorcountislessthen41: TMenuItem;
    N6cylindercars1: TMenuItem;
    Pricemorethen1000001: TMenuItem;
    Customannotationssettings1: TMenuItem;
    procedure FormCreate(Sender: TObject);
    procedure actCustomAnnotationSettingsExecute(Sender: TObject);
    procedure actDoorCountExecute(Sender: TObject);
    procedure actScrollAnnotationsActiveExecute(Sender: TObject);
    procedure actShowErrorsExecute(Sender: TObject);
    procedure actShowSearchResultsExecute(Sender: TObject);
    procedure actShowFocusedRowExecute(Sender: TObject);
    procedure cxDBVerticalGrid1GetScrollbarAnnotationHint(Sender: TObject; AAnnotationRowIndexLists: TdxScrollbarAnnotationRowIndexLists; var AHint: string);
    procedure cxDBVerticalGrid1KeyDown(Sender: TObject; var Key: Word; Shift: TShiftState);
    procedure cxDBVerticalGrid1MPGCityPropertiesValidateDrawValue(
      Sender: TcxCustomEditorRowProperties; ARecordIndex: Integer;
      const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxDBVerticalGrid1PopulateCustomScrollbarAnnotationRowIndexList(Sender: TObject; AAnnotationIndex: Integer; ARowIndexList: TdxScrollbarAnnotationRowIndexList);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  ScrollbarAnnotationsDemoMainForm: TScrollbarAnnotationsDemoMainForm;

implementation

{$R *.dfm}

uses
  Math, StrUtils, CarsData, CustomAnnotationSettings;

procedure TScrollbarAnnotationsDemoMainForm.FormCreate(Sender: TObject);
begin
  inherited;
  SCustomAnnotationstrs[0] := '2-Door Cars';
  SCustomAnnotationstrs[1] := 'Price is Greater Than ' + CurrToStrF(100000, ffCurrency, 0);
  SCustomAnnotationstrs[2] := '6-Cylinder Cars';
  cxDBVerticalGrid1.DataController.FindCriteria.Text := 'V8';
  actDoorCount.Caption := SCustomAnnotationstrs[0];
  actPrice.Caption := SCustomAnnotationstrs[1];
  actCylinderCount.Caption := SCustomAnnotationstrs[2];
end;

procedure TScrollbarAnnotationsDemoMainForm.actCustomAnnotationSettingsExecute(Sender: TObject);
begin
  if frmCustomAnnotationSettings = nil then
  begin
    Application.CreateForm(TfrmCustomAnnotationSettings, frmCustomAnnotationSettings);
    frmCustomAnnotationSettings.Initialize(cxDBVerticalGrid1.ScrollbarAnnotations.CustomAnnotations);
  end;
  frmCustomAnnotationSettings.Show;
end;

procedure TScrollbarAnnotationsDemoMainForm.actDoorCountExecute(Sender: TObject);
begin
  cxDBVerticalGrid1.ScrollbarAnnotations.CustomAnnotations[(Sender as TAction).Tag].Visible := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actScrollAnnotationsActiveExecute(Sender: TObject);
begin
  cxDBVerticalGrid1.ScrollbarAnnotations.Active := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actShowErrorsExecute(Sender: TObject);
begin
  cxDBVerticalGrid1.ScrollbarAnnotations.ShowErrors := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actShowSearchResultsExecute(Sender: TObject);
begin
  cxDBVerticalGrid1.ScrollbarAnnotations.ShowSearchResults := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actShowFocusedRowExecute(Sender: TObject);
begin
  cxDBVerticalGrid1.ScrollbarAnnotations.ShowFocusedRow := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.cxDBVerticalGrid1GetScrollbarAnnotationHint(Sender: TObject; AAnnotationRowIndexLists: TdxScrollbarAnnotationRowIndexLists; var AHint: string);

  function GetValue(ARecordIndex: Integer; AColumn: TcxDBEditorRow): Variant;
  begin
    Result := AColumn.Properties.Values[ARecordIndex];
  end;

var
  AKind: TdxScrollbarAnnotationKind;
  AList: TdxScrollbarAnnotationRowIndexList;
  I: Integer;
  ACount: Integer;
  AAnnotationHint: string;
begin
  for AKind in AAnnotationRowIndexLists.Keys do
  begin
    AList := AAnnotationRowIndexLists[AKind];
    ACount := Min(5, AList.Count);
    for I := 0 to ACount - 1 do
    begin
      case AKind of
        0:
          begin
            AAnnotationHint := Format(' %s doors', [IntToStr(GetValue(AList[I], clDoorCount))]);
          end;
        1:
          begin
            AAnnotationHint := Format('(price: %s)', [FormatFloat('$,0;($,0)', GetValue(AList[I], clPrice))]);
          end;
        2:
          begin
            AAnnotationHint := Format(' %s cylinders', [IntToStr(GetValue(AList[I], clCilinderCount))]);
          end;
        dxErrorScrollbarAnnotationID:
          begin
            AAnnotationHint := ' - MPG City is not specified!';
          end;
      else
        AAnnotationHint := '';
      end;
      if AAnnotationHint <> '' then
      begin
        AAnnotationHint := Format('%s %s', [GetValue(AList[I], clName), GetValue(AList[I], clModification)]) + AAnnotationHint;
        AHint := AHint + IfThen(AHint <> '', dxCRLF) + AAnnotationHint;
      end;
    end;
    if AList.Count > ACount then
      AHint := AHint + dxCRLF + Format('And %d more records...', [AList.Count - ACount]);
  end;
end;

procedure TScrollbarAnnotationsDemoMainForm.cxDBVerticalGrid1KeyDown(Sender: TObject; var Key: Word; Shift: TShiftState);
begin
  if Key = VK_F4 then
    cxDBVerticalGrid1.GoToNextScrollbarAnnotation([dxErrorScrollbarAnnotationID], not (ssShift in Shift))
end;

procedure TScrollbarAnnotationsDemoMainForm.cxDBVerticalGrid1MPGCityPropertiesValidateDrawValue(
  Sender: TcxCustomEditorRowProperties; ARecordIndex: Integer;
  const AValue: Variant; AData: TcxEditValidateInfo);
begin
  if VarIsStr(AValue) and (AValue = '') or (AValue = Null) then
    AData.ErrorType := eetError;
end;

procedure TScrollbarAnnotationsDemoMainForm.cxDBVerticalGrid1PopulateCustomScrollbarAnnotationRowIndexList(Sender: TObject; AAnnotationIndex: Integer; ARowIndexList: TdxScrollbarAnnotationRowIndexList);

  function GetValue(ARecordIndex: Integer; AColumn: TcxDBEditorRow): Variant;
  begin
    Result := AColumn.Properties.Values[ARecordIndex];
  end;

var
  I: Integer;
  ACondition: Boolean;
  AValue: Variant;
begin
  for I := 0 to cxDBVerticalGrid1.RecordCount - 1 do
  begin
    case AAnnotationIndex of
      0:
        begin
          AValue := GetValue(I, clDoorCount);
          ACondition := not VarIsNull(AValue) and (AValue = 2);
        end;
      1:
        begin
          ACondition := GetValue(I, clPrice) > 100000;
        end;
      2:
        ACondition := GetValue(I, clCilinderCount) = 6
    else
      ACondition := False;
    end;
    if ACondition then
      ARowIndexList.Add(I);
  end;
end;

end.

File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressVerticalGrid/Demos/CBuilder/ScrollbarAnnotationsDemo/ScrollbarAnnotationsDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ScrollbarAnnotationsDemoMain.h"
#include "CarsData.h"
#include "CustomAnnotationSettings.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxControls"
#pragma link "cxDBVGrid"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxInplaceContainer"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxStyles"
#pragma link "cxTextEdit"
#pragma link "cxVGrid"
#pragma resource "*.dfm"
TScrollbarAnnotationsDemoMainForm *ScrollbarAnnotationsDemoMainForm;
String SCustomAnnotationStrs[3];
//---------------------------------------------------------------------------
__fastcall TScrollbarAnnotationsDemoMainForm::TScrollbarAnnotationsDemoMainForm(TComponent* Owner)
	: TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------
Variant TScrollbarAnnotationsDemoMainForm::GetValue(int ARecordIndex, TcxDBEditorRow *AColumn)
{
  return AColumn->Properties->Values[ARecordIndex];
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actScrollAnnotationsActiveExecute(TObject *Sender)

{
  cxDBVerticalGrid1->ScrollbarAnnotations->Active = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actDoorCountExecute(TObject *Sender)
{
  cxDBVerticalGrid1->ScrollbarAnnotations->CustomAnnotations->Items[dynamic_cast<TAction *>(Sender)->Tag]->Visible = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actShowErrorsExecute(TObject *Sender)
{
  cxDBVerticalGrid1->ScrollbarAnnotations->ShowErrors = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actShowSearchResultsExecute(TObject *Sender)

{
  cxDBVerticalGrid1->ScrollbarAnnotations->ShowSearchResults = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actShowFocusedRowExecute(TObject *Sender)
{
  cxDBVerticalGrid1->ScrollbarAnnotations->ShowFocusedRow = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actCustomAnnotationSettingsExecute(TObject *Sender)

{
  if (!frmCustomAnnotationSettings)
  {
	Application->CreateForm(__classid(TfrmCustomAnnotationSettings), &frmCustomAnnotationSettings);
	frmCustomAnnotationSettings->Initialize(cxDBVerticalGrid1->ScrollbarAnnotations->CustomAnnotations);
  };
  frmCustomAnnotationSettings->Show();
}
//---------------------------------------------------------------------------

void __fastcall TScrollbarAnnotationsDemoMainForm::cxDBVerticalGrid1PopulateCustomScrollbarAnnotationRowIndexList(TObject *Sender,
		  int AAnnotationIndex, TdxScrollbarAnnotationRowIndexList *ARowIndexList)
{
  bool ACondition = false;
  Variant AValue;
  for (int i = 0; i < cxDBVerticalGrid1->RecordCount; i++)
  {
	switch (AAnnotationIndex)
	{
	  case 0:
		  AValue = GetValue(i, clDoorCount);
		  ACondition = !VarIsNull(AValue) && (AValue == 2);
		  break;
	  case 1:
		ACondition = (GetValue(i, clPrice) > 100000);
		break;
	  case 2:
		ACondition = (GetValue(i, clCilinderCount) == 6);
		break;
	  default:
	    ACondition = false;
    };
	if (ACondition)
	  ARowIndexList->Add(i);
  };
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::cxDBVerticalGrid1GetScrollbarAnnotationHint(TObject *Sender,
		  TdxScrollbarAnnotationRowIndexLists *AAnnotationRowIndexLists, UnicodeString &AHint)

{
  System::DynamicArray<TdxScrollbarAnnotationKind> AKinds = AAnnotationRowIndexLists->Keys->ToArray();
  TdxScrollbarAnnotationKind AKind;
  int ACount;
  String AAnnotationHint;
  TdxScrollbarAnnotationRowIndexList* AList;
  int ARecordIndex;
  for (int i = 0; i < AKinds.Length; i++)
  {
	AKind = AKinds[i];
	AAnnotationRowIndexLists->TryGetValue(AKind, AList);
	ACount = Math::Min(5, AList->Count);
	for (int j = 0; j < ACount; j++)
	{
	  ARecordIndex = AList->Items[j];
	  switch (AKind)
	  {
		case 0:
			AAnnotationHint = String::Format(" %s doors", ARRAYOFCONST((GetValue(ARecordIndex, clDoorCount))));
			break;
		case 1:
		   {
			float f = GetValue(ARecordIndex, clPrice);
			AAnnotationHint = " (price: " + FormatFloat("$,0;($,0)", f) + ")";
			break;
		   }
		case 2:
			AAnnotationHint = String::Format(" %s cylinders", ARRAYOFCONST((GetValue(ARecordIndex, clCilinderCount))));
			break;
		case dxErrorScrollbarAnnotationID:
			AAnnotationHint = " - MPG City is not specified!";
			break;
		default:
		    AAnnotationHint = "";
	  };
	  if (AAnnotationHint != "")
	  {
		AAnnotationHint = String::Format("%s %s%s", ARRAYOFCONST((GetValue(ARecordIndex, clName), GetValue(ARecordIndex, clModification), AAnnotationHint)));
		AHint = AHint + IfThen(AHint != "", dxCRLF) + AAnnotationHint;
	  };
	};
	if (AList->Count > ACount)
	  AHint = AHint + dxCRLF + "And " + IntToStr(AList->Count - ACount) +  " more records...";
  };
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::cxDBVerticalGrid1KeyDown(TObject *Sender, WORD &Key,
          TShiftState Shift)
{
  Set <TdxScrollbarAnnotationKind, 0, 255> AKinds;
  AKinds << dxErrorScrollbarAnnotationID;
  if (Key == VK_F4)
	cxDBVerticalGrid1->GoToNextScrollbarAnnotation(AKinds, !Shift.Contains(ssShift));
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::cxDBVerticalGrid1MPGCityPropertiesValidateDrawValue(TcxCustomEditorRowProperties *Sender,
          int ARecordIndex, const Variant &AValue, TcxEditValidateInfo *AData)

{
  if (VarIsNull(AValue))
	AData->ErrorType = eetError;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::FormCreate(TObject *Sender)
{
  SCustomAnnotationStrs[0] = "2-Door Cars";
  SCustomAnnotationStrs[1] = "Price is Greater Than " + CurrToStrF(100000, ffCurrency, 0);
  SCustomAnnotationStrs[2] = "6-Cylinder Cars";
  cxDBVerticalGrid1->DataController->FindCriteria->Text = "V8";
  actDoorCount->Caption = SCustomAnnotationStrs[0];
  actPrice->Caption = SCustomAnnotationStrs[1];
  actCylinderCount->Caption = SCustomAnnotationStrs[2];
}
//---------------------------------------------------------------------------

